src/spi.rs

```diff
@@ -1266,41 +1349,165 @@ impl Drop for Lock {
     }
 }
 
-// These parameters assume full duplex.
-fn polling_transmit(
-    handle: spi_device_handle_t,
-    read: *mut u8,
-    write: *const u8,
-    transaction_length: usize,
-    rx_length: usize,
-    _keep_cs_active: bool,
-) -> Result<(), EspError> {
-    #[cfg(esp_idf_version = "4.3")]
-    let flags = 0;
-
-    // This unfortunately means that this implementation is incorrect for esp-idf < 4.4.
-    // The CS pin should be kept active through transactions.
-    #[cfg(not(esp_idf_version = "4.3"))]
-    let flags = if _keep_cs_active {
-        SPI_TRANS_CS_KEEP_ACTIVE
-    } else {
-        0
-    };
+static VTABLE: RawWakerVTable = RawWakerVTable::new(
+    |_| RawWaker::new(core::ptr::null(), &VTABLE),
+    |_| {},
+    |_| {},
+    |_| {},
+);
 
-    let mut transaction = spi_transaction_t {
-        flags,
-        __bindgen_anon_1: spi_transaction_t__bindgen_ty_1 {
-            tx_buffer: write as *const _,
-        },
-        __bindgen_anon_2: spi_transaction_t__bindgen_ty_2 {
-            rx_buffer: read as *mut _,
-        },
-        length: (transaction_length * 8) as _,
-        rxlength: (rx_length * 8) as _,
-        ..Default::default()
-    };
+enum Transmit {
+    Blocking(bool),
+    Async(u8),
+}
+
+impl Transmit {
+    // These parameters assume full duplex.
+    pub async fn transmit(
+        &self,
+        handle: spi_device_handle_t,
+        read: *mut u8,
+        write: *const u8,
+        transaction_length: usize,
+        rx_length: usize,
+        keep_cs_active: bool,
+    ) -> Result<(), EspError> {
+        let mut transaction = Self::create_transaction(
+            handle,
+            read,
+            write,
+            transaction_length,
+            rx_length,
+            keep_cs_active,
+        )?;
+
+        match self {
+            Self::Async(host) => {
+                loop {
+                    match esp!(unsafe {
+                        spi_device_queue_trans(handle, &mut transaction as *mut _, delay::NON_BLOCK)
+                    }) {
+                        Ok(_) => break,
+                        Err(e) if e.code() != ESP_ERR_TIMEOUT => return Err(e),
+                        _ => NOTIFIER[*host as usize].wait().await,
+                    }
+                }
+
+                Completion::new(
+                    async {
+                        loop {
```

Comment:
Fair enough. Besides "just in case" also because all devices attached to the same host/bus/hardware currently share a single `Notification` object, hence all of the devices competing (and waiting) for the next transaction to start/complete would be awaken even if a transaction for a completely separate device was signalled as started/completed - but obviously not all of them will be able to send/receive - only one, as the hardware is shared.

With that said, I think you have a point though - I should switch to a per-device Notification object. This would allow me to get rid of the `MultiNotification` thing as well, as I'll have a single waker per notification then (the waker for that particular device).
 
The above wont't mean I can completely get rid of the loop - as I currently don't care if I'm notified about a transaction being started or completed - it is the same notification. And the loop takes care of that.

Reply:
> all devices attached to the same host/bus/hardware currently share a single `Notification` object

Ah, I didn't realised this. So this code behaves like a condition variable loop basically, cool.

Let's see what you're able to pull of with the user data then as that'll help with the per-device stuff.

Reply:
Just an update that the `Notification` object is no longer shared across multiple devices. In fact, it is now specific for the transaction which is currently running.